В цьому ДЗ потренуємось виконувати Z-тести для тестування гіпотез в різних задачах. Для коректної імплементації та інтерпретації тестів рекоемндую переглянути спершу документацію:

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html

### Завдання 1: Z-тест для однієї вибірки

**Мета**: Перевірити, чи середнє значення вибірки відрізняється від гіпотетичного середнього.

**Припущення**: Середній ріст студентів університету вважається 170 см. Ми хочемо перевірити, чи середній ріст випадкової вибірки студентів істотно більший. Вибірка задана в змінній `heights`.

Для виконання завдання імпортуйте метод для проведення `ztest`, виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення про середній зріст студентів на рівні значущості 0.05?

Використовуємо тут Z-тест, оскільки вибірка достатньо велика (більше 30 екземплярів) і можемо вважати, що її стандартне відхилення близьке до ст. відхилення генеральної сукупності, а значить - відоме.

In [87]:
import numpy as np
from scipy.stats import ttest_ind
from scipy.stats import ttest_rel
from statsmodels.stats.weightstats import ztest

In [55]:
heights = [174, 171, 175, 179, 170, 170, 179, 175, 169, 174, 169, 169, 173, 162, 163, 169, 166,
           173, 167, 164, 179, 170, 172, 164, 169, 175, 169, 174, 169, 169, 173, 162, 177]

In [56]:
mean_heights = np.mean(heights)
alpha = 0.05
null_mean = 170

In [57]:
mean_heights

170.6969696969697

In [61]:
ztest_Score, p_value = ztest(heights, value = null_mean, alternative='larger')

In [62]:
ztest_Score, p_value

(0.8482235371587619, 0.19815674571091746)

In [63]:
if(p_value <  alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Н0 не може бути відхилена.


Ми не можемо відхилити Н0 та стверджувати, що середній зріст випадкової вибірки вищий за 170, за результатами тесту середній зріст дорівнює або меньше 170 (бо така Н0). Хоча середнє вибірки у нас 170.6969696969697, але перевіримо за допомогою тесту чи середній зріст вибірки меньший за 170

In [64]:
ztest_Score, p_value = ztest(heights, value = null_mean, alternative='smaller')
print(ztest_Score, p_value)
if(p_value <  alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

0.8482235371587619 0.8018432542890825
Н0 не може бути відхилена.


За результатами ліво та правостороннього тестів бачимо, що середній зріст вибірки не більший і не меньший за 170. Можна було також одразу провести alternative='two-sided', щоб спростувати чи підтвердити, що середнє по вибірці та null_mean(170) однакові.

### Завдання 2: Z-тест для двох незалежних вибірок

**Завдання 2.1.**

**Мета**: Перевірити, чи існує статистична різниця між середніми оцінками двох груп студентів.

**Припущення**: Група A отримала новий навчальний курс, а група B продовжувала зі стандартним курсом. Ми перевіряємо, чи є новий курс ефективнішим.

Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення щодо студентів на рівні значущості 0.05?

In [65]:
group_a_scores = [78.55, 72.25, 79.88, 75.  , 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.9 , 79.04, 68.2 , 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.7 , 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.6 ,
       77.07, 72.47, 72.02, 82.06]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19]

In [66]:
np.mean(group_a_scores), np.mean(group_b_scores)

(77.08125, 74.73875000000001)

In [67]:
ztest_Score, p_value = ztest(group_a_scores, group_b_scores, alternative='two-sided')
ztest_Score, p_value

(2.357354864751502, 0.01840564855252801)

In [68]:
# alpha = 0.05
if(p_value <  alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Відхилити Н0.


Бачимо, що за результатом двостороннього тесту ми спостерігаємо різницю між двома вибірками. Бачимо, що p_value <  alpha, а отже є статистична значущисть між двома курсами. Тепер варто перевірити, який курс дає кращі результати. Для цього ми можемо порівняти середні значення двох вибірок за умови, що ми вже розуміємо, що є статистично значуща різниця між ними.

In [70]:
    if np.mean(group_a_scores) > np.mean(group_b_scores):
        print("Новий курс (група A) ефективніший за стандартний (група B).")
    else:
        print("Стандартний курс (група B) кращий за новий (група A).")

Новий курс (група A) ефективніший за стандартний (група B).


**Завдання 2.2.**


**Мета**: Побачити, як впливає розмір вибірок на результат тесту.

**Завдання**: Уявіть, що з групи А ми маємо тільки перші 5 записів, а з групи В - усі. Таке могло статись, якщо ми вже маємо результати тестів учнів за попередньою програмою, а тести за новою програмою здали наразі лише 5 учнів і ми вирішили не чекати довше.
Виберіть перші 5 записів для групи А і проведіть t-test (для цього використовуємо наступний [метод](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html)). Виведіть значення p-value. Зробіть висновок, який результат ви бачите щодо стат. значущості на рівні 0.05 для цього експерименту?

Тут проводимо t-test оскільки одна з вибірок - дуже мала.

In [73]:
group_a_small = group_a_scores[:5]
group_a_small, np.mean(group_a_small)

([78.55, 72.25, 79.88, 75.0, 76.54], 76.444)

In [78]:
t_stat, p_value = ttest_ind(group_a_small, group_b_scores, equal_var=False)

print(t_stat, p_value)

if p_value < 0.05:
    print("Різниця між групами є статистично значущою.")
else:
    print("Різниця між групами не є статистично значущою.")

1.1226527301193456 0.3010736990344312
Різниця між групами не є статистично значущою.


За результатами тесту бачимо, що немає статистичної значущості між двома вибірками. Отже бачимо, що недостатня кількість даних значно вплинула на результати


### Завдання 3\*: Z-тест для двох **пов'язаних** вибірок

**Мета**: Перевірити, чи вплинув тренінг на продуктивність працівників, коли всі справробітники проходили тренінг.

**Припущення**: Наші працівники проходять однаковий тренінг, і ми хочемо з'ясувати, чи покращилися їхні показники після тренінгу на рівні значущості 0.05. Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи поліпшились показники працівників на рівні значущості 0.05?

Зверніть увагу, ці вибірки повʼязані між собою, значить не буде коректно виконати z-test між ними, натомість, ми можемо порівняти, чи різниця в результатах (продуктивності співробітників) є стат значущою.

In [80]:
before_training = [57.82, 37.63, 36.8 , 55.22, 52.97, 52.5 , 53.46, 43.2 , 52.32,
       52.93, 42.86, 68.66, 54.74, 38.09, 56.57, 40.25, 57.87, 61.59,
       41.79, 59.63, 54.13, 58.22, 68.97, 47.55, 42.46, 41.1 , 41.84,
       49.23, 53.41, 52.77]

after_training = [62.47, 40.66, 42.7 , 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
       55.48, 47.28, 72.6 , 57.59, 39.39, 56.54, 42.36, 62.58, 65.01,
       42.3 , 62.98, 57.9 , 59.45, 72.28, 50.66, 43.18, 44.82, 45.96,
       54.4 , 58.52, 53.01]

In [81]:
np.mean(before_training), np.mean(after_training)

(50.885999999999996, 54.03600000000001)

In [84]:
ztest_Score, p_value = ztest(before_training, after_training, alternative='two-sided')
ztest_Score, p_value

(-1.3668871347639466, 0.17166068512470822)

In [85]:
if p_value < 0.05:
    print("Різниця між групами є статистично значущою.")
else:
    print("Різниця між групами не є статистично значущою.")

Різниця між групами не є статистично значущою.


За результатими z-тесту бачимо, що немає статистичної знаущості між вибірками. Спробуємо провести t-test, оскільки тут вибірки повʼязані - https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_rel.html#ttest-rel.

In [89]:
t_stat, p_value = ttest_rel(before_training, after_training)

print("t-Statistic:", t_stat)
print("p-Value:", p_value)

alpha = 0.05
if p_value < alpha:
    print("Різниця між групами є статистично значущою.")
else:
    print("Різниця між групами не є статистично значущою.")

t-Statistic: -9.13890009623916
p-Value: 4.885873625502449e-10
Різниця між групами є статистично значущою.


За результатами t-тесту бачимо, що тренінг вплинув на ефективність працівників. Якщо порівняти середні значення до 50.885999999999996 та після 54.03600000000001 тренінгу, то робимо висновок, що тренінг позитивно вплинув на продуктивність.